From 046a0e146cfc1121879d8427ad239d77f456eeec Mon Sep 17 00:00:00 2001
From: GuaiYiHu <guaiyihu@foxmail.com>
Date: Sun, 11 Oct 2026 18:59:35 +0800
Subject: [PATCH] Bluetooth: Guard LE advertising and feature access for
 classic controllers

Adapted from commit 026e526c317034adbf353ffc9eace95881b69767. Android 12 uses a typed adapter-properties callback; return zeroed LE capabilities without querying a classic-only controller.

Change-Id: I942c75a113dc61681acd5225fb6c993c83e5bceb
---
 bta/dm/bta_dm_act.cc | 2 ++
 btif/src/btif_dm.cc  | 6 +++++-
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/bta/dm/bta_dm_act.cc b/bta/dm/bta_dm_act.cc
index 689f81fa1..fd5a77144 100644
--- a/bta/dm/bta_dm_act.cc
+++ b/bta/dm/bta_dm_act.cc
@@ -386,6 +386,7 @@ void BTA_dm_on_hw_on() {
 
   BTM_WritePageTimeout(p_bta_dm_cfg->page_timeout);
 
+  if (controller_get_interface()->supports_ble()) {
 #if (BLE_VND_INCLUDED == TRUE)
   BTM_BleReadControllerFeatures(bta_dm_ctrl_features_rd_cmpl_cback);
 #else
@@ -394,6 +395,7 @@ void BTA_dm_on_hw_on() {
    * advertising here */
   btm_ble_adv_init();
 #endif
+  }
 
   /* Earlier, we used to invoke BTM_ReadLocalAddr which was just copying the
      bd_addr
diff --git a/btif/src/btif_dm.cc b/btif/src/btif_dm.cc
index 61892ece6..ea4623ee9 100644
--- a/btif/src/btif_dm.cc
+++ b/btif/src/btif_dm.cc
@@ -1697,11 +1697,15 @@ static void btif_dm_upstreams_evt(uint16_t event, char* p_param) {
 
     case BTA_DM_LE_FEATURES_READ: {
       tBTM_BLE_VSC_CB cmn_vsc_cb;
-      bt_local_le_features_t local_le_features;
+      bt_local_le_features_t local_le_features = {};
       bt_property_t prop;
       prop.type = BT_PROPERTY_LOCAL_LE_FEATURES;
       prop.val = (void*)&local_le_features;
       prop.len = sizeof(bt_local_le_features_t);
+      if (!controller_get_interface()->supports_ble()) {
+        invoke_adapter_properties_cb(BT_STATUS_SUCCESS, 1, &prop);
+        break;
+      }
 
       /* LE features are not stored in storage. Should be retrived from stack */
       BTM_BleGetVendorCapabilities(&cmn_vsc_cb);
